motoko/icrc2-swap/src/swap/main.mo

```diff
@@ -0,0 +1,294 @@
+    // correctly deduct it from their balance.
+    let fee = switch (args.fee) {
+      case (?f) { f };
+      case (null) { await token.icrc1_fee() };
+    };
+
+    // Perform the transfer, to capture the tokens.
+    let transfer_result = await token.icrc2_transfer_from({
+      spender_subaccount = args.spender_subaccount;
+      from = args.from;
+      to = { owner = Principal.fromActor(this); subaccount = null };
+      amount = args.amount;
+      fee = ?fee;
+      memo = args.memo;
+      created_at_time = args.created_at_time;
+    });
+
+    // Check that the transfer was successful.
+    let block_height = switch (transfer_result) {
+      case (#Ok(block_height)) { block_height };
+      case (#Err(err)) {
+        // Transfer failed. There's no cleanup for us to do, so we can just
+        // wrap and return the error to the frontend.
+        return #err(#TransferFromError(err));
+      };
+    };
+
+    // From here on out, we need to make sure that this function does *not*
+    // fail. If it failed, the token transfer would be complete (meaning we
+    // would have the user's tokens), but we would not have credited their
+    // account yet, so this canister would not *know* that it had received the
+    // user's tokens.
+    // 
+    // If the function *can* fail here after this point, we should either:
+    // - Move that code to a separate action later
+    // - Have failure-handling code which returns the user's tokens
+    
+    // Credit the sender's account
+    let sender = args.from.owner;
+    let old_balance = Option.get(balances.get(sender), 0 : Nat);
+    balances.put(sender, old_balance + args.amount);
+
+    // Return the "block height" of the transfer
+    #ok(block_height)
+  };
+
+  public type SwapArgs = {
+    user_a: Principal;
+    user_b: Principal;
+  };
+
+  public type SwapError = {
+    // Left as a placeholder for future implementors, in case their swap
+    // function could fail.
+  };
+
+  // Swap TokenA for TokenB
+  // - Exchange tokens between the two given users.
+  // - For this example, there will be no clever swap mechanism, it simply swaps all
+  //   deposits for the two users. Designing a useful and safer swap mechanism
+  //   is left as an exercise for the reader.
+  // - UserA's full balance of TokenA is given to UserB, and UserB's full
+  //   balance of TokenB is given to UserA.
+  public shared(msg) func swap(args: SwapArgs): async Result.Result<(), SwapError> {
+    // Because both tokens were deposited before calling swap, we can execute
+    // this function atomically. To do that there must be no `await` calls in
+    // this function. Additionally, we need to be careful with the order that
+    // we update any internal state variables. If this function were to update
+    // some state variables, then fail, before updating others, it could leave
+    // this canister with inconsistent internal state.
```

Comment:
Wrong. This is only a concern if you `await` in the middle of state changes. If this code traps anywhere all state changes will be reverted

Reply:
Ah true. 🤔 Let me rewrite this...